air/src/aux_table/bitwise_pow2/bitwise.rs

```diff
@@ -0,0 +1,441 @@
+    fn bitwise_op_flag(&self) -> E;
+    /// The selector flag for the bitwise AND operation.
+    fn bitwise_and_flag(&self) -> E;
+    /// The selector flag for the bitwise OR operation.
+    fn bitwise_or_flag(&self) -> E;
+    /// The selector flag for the bitwise XOR operation.
+    fn bitwise_xor_flag(&self) -> E;
+}
+
+impl<E: FieldElement> EvaluationFrameExt<E> for &EvaluationFrame<E> {
+    // --- Column accessors -----------------------------------------------------------------------
+
+    fn selector(&self, index: usize) -> E {
+        self.current()[SELECTOR_COL_RANGE.start + index]
+    }
+    fn a(&self) -> E {
+        self.current()[A_COL_IDX]
+    }
+    fn a_next(&self) -> E {
+        self.next()[A_COL_IDX]
+    }
+    fn a_bit(&self, index: usize) -> E {
+        self.current()[A_COL_RANGE.start + index]
+    }
+    fn b(&self) -> E {
+        self.current()[B_COL_IDX]
+    }
+    fn b_next(&self) -> E {
+        self.next()[B_COL_IDX]
+    }
+    fn b_bit(&self, index: usize) -> E {
+        self.current()[B_COL_RANGE.start + index]
+    }
+    fn bit_decomp(&self) -> &[E] {
+        &self.current()[A_COL_RANGE.start..B_COL_RANGE.end]
+    }
+    fn bit_decomp_next(&self) -> &[E] {
+        &self.next()[A_COL_RANGE.start..B_COL_RANGE.end]
+    }
+    fn output(&self) -> E {
+        self.current()[OUTPUT_COL_IDX]
+    }
+    fn output_next(&self) -> E {
+        self.next()[OUTPUT_COL_IDX]
+    }
+
+    // --- Intermediate variables & helpers -------------------------------------------------------
+    fn a_agg_bits(&self) -> E {
+        agg_bits(self.current(), A_COL_RANGE.start)
+    }
+    fn a_agg_bits_next(&self) -> E {
+        agg_bits(self.next(), A_COL_RANGE.start)
+    }
+    fn b_agg_bits(&self) -> E {
+        agg_bits(self.current(), B_COL_RANGE.start)
+    }
+    fn b_agg_bits_next(&self) -> E {
+        agg_bits(self.next(), B_COL_RANGE.start)
+    }
+
+    // --- Flags ----------------------------------------------------------------------------------
+
+    fn bitwise_op_flag(&self) -> E {
+        binary_not(self.selector(0) * self.selector(1))
+    }
+    fn bitwise_and_flag(&self) -> E {
+        binary_not(self.selector(0)) * binary_not(self.selector(1))
+    }
+    fn bitwise_or_flag(&self) -> E {
+        binary_not(self.selector(0)) * self.selector(1)
+    }
+    fn bitwise_xor_flag(&self) -> E {
+        self.selector(0) * binary_not(self.selector(1))
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+/// Aggregate 4 decomposed bits representing a 4-bit binary value into a decimal value, starting
+/// from `start_idx` in the provided row.
+fn agg_bits<E: FieldElement>(row: &[E], start_idx: usize) -> E {
+    let mut result = E::ZERO;
+    for bit_idx in 0..NUM_DECOMP_BITS {
+        result += E::from(2_u64.pow(bit_idx as u32)) * row[start_idx + bit_idx];
+    }
```

Comment:
There is one potential optimization here, though it would be difficult to do in the current structure: we are multiplying by a small power of two and then summing up the results - thus, in theory, we could just aggregate results in a 128-bit integer and perform only a single reduction in the end.

This works only when we are in the base field - but since we have a generic parameter `E`, I'm not quite sure how to do it. So, maybe for now just leave a note saying that there is potential optimization here?

Reply:
I hadn't considered this, but it's a nice point. I've added a comment.